sw/device/lib/crypto/impl/kdf.c

```diff
@@ -15,13 +15,197 @@
+    // The underlying HMAC implementation is not currently hardened.
+    return OTCRYPTO_NOT_IMPLEMENTED;
+  }
+
+  // Check for null label with nonzero length.
+  if (kdf_label.data == NULL && kdf_label.len != 0) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check for null context with nonzero length.
+  if (kdf_context.data == NULL && kdf_context.len != 0) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check the private key checksum.
+  if (integrity_blinded_key_check(&key_derivation_key) != kHardenedBoolTrue) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check non-zero length for keying_material.
+  if (required_byte_len == 0) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Infer the digest size.
+  size_t digest_word_len = 0;
+  HARDENED_TRY(digest_num_words_from_key_mode(
+      key_derivation_key.config.key_mode, &digest_word_len));
+
+  // Ensure that the derived key is a symmetric key masked with XOR and is not
+  // supposed to be hardware-backed.
+  HARDENED_TRY(keyblob_ensure_xor_masked(keying_material->config));
+
+  // Check `keying_material` key length.
+  if (keying_material->config.hw_backed == kHardenedBoolTrue) {
+    // The case where `keying_material` is hw_backed is addressed by
+    // `otcrypto_hw_backed_key` function in `key_transport.h`.
+    return OTCRYPTO_BAD_ARGS;
+  } else if (keying_material->config.hw_backed == kHardenedBoolFalse) {
+    if (keying_material->config.key_length != required_byte_len ||
+        keying_material->keyblob_length !=
+            keyblob_num_words(keying_material->config) * sizeof(uint32_t)) {
+      return OTCRYPTO_BAD_ARGS;
+    }
+  } else {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check that the unmasked key length is not too large for HMAC CTR
+  // (see NIST SP 800-108r1, section 4.1)
+  size_t required_word_len = ceil_div(required_byte_len, sizeof(uint32_t));
+  size_t num_iterations = ceil_div(required_word_len, digest_word_len);
+  if (launder32(num_iterations) > UINT32_MAX) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+  HARDENED_CHECK_LE(num_iterations, UINT32_MAX);
+
+  // Check if label or context contain 0x00 bytes
+  // Since 0x00 is used as the delimiter between label and context
+  // there shouldn't be multiple instances of them in input data
+  HARDENED_TRY(check_zero_byte(kdf_label));
+  HARDENED_TRY(check_zero_byte(kdf_context));
+
+  // Create a buffer that holds input data:
+  // [i]_2 || Label || 0x00 || Context || [L]_2 (see NIST SP 800-108r1,
+  // section 4.1)
+  // The counter value is updated within the loop instead of here
+
+  uint8_t input_data[4 + kdf_label.len + 1 + kdf_context.len + 4];
+  memcpy(input_data + 4, kdf_label.data, kdf_label.len);
+  input_data[4 + kdf_label.len] = 0x00;
+  memcpy(input_data + 4 + kdf_label.len + 1, kdf_context.data, kdf_context.len);
+  uint8_t required_len_bytestr[4];
+  uint_to_bytes(required_byte_len * 8, required_len_bytestr);
+  memcpy(input_data + 4 + kdf_label.len + 1 + kdf_context.len,
+         required_len_bytestr, 4);
```

Comment:
nit: I find this bit kind of confusing to read with all the `+ 1` and `+ 4` terms. I also don't see where the first 4 bytes are getting initialized (are they zero?) But ultimately, I don't think you really need to do this copying at all -- I suggest `otcrypto_hmac_update` instead:
```c
// Setup:
uint32_t counter = 0;
uint8_t zero = 0x00;
uint32_t required_bit_len = __builtin_bswap32(required_byte_len * 8);

// In the loop:
otcrypto_hmac_context_t ctx;
HARDENED_TRY(otcrypto_hmac_init(&ctx, &key_derivation_key));
uint32_t counter_be = __builtin_bswap32(counter);
HARDENED_TRY(otcrypto_hmac_update(&ctx, (otcrypto_const_byte_buf_t) { .data = &counter_be, .size = sizeof(counter_be) }));
HARDENED_TRY(otcrypto_hmac_update(&ctx, kdf_label));
HARDENED_TRY(otcrypto_hmac_update(&ctx, (otcrypto_const_byte_buf_t) { .data = &zero, .size = sizeof(zero) }));
HARDENED_TRY(otcrypto_hmac_update(&ctx, (otcrypto_const_byte_buf_t) { .data = &required_bit_len, .size = sizeof(required_bit_len) }));
uint32_t *tag_dest = keying_material_data + i * digest_word_len;
HARDENED_TRY(otcrypto_hmac_final(&ctx, (otcrypto_word32_buf_t) { .data = tag_dest, .size = digest_word_len }));
```

Reply:
Thanks you for the detailed correction. I substituted in the provided code, and I will send Moritz the updated version shortly.